### DISCUSSION TITLE
Custom field in custom script does not save to database

### DISCUSSION THREAD

**Discussion by MalfuncEddie** [netbox-community/netbox#9672] [Q&A] [answered]
Hi,

I am kind of stuck on why the following piece of custom script code does not update the  VirtualMachine.
When running I, I get the following log loglines

<html>
<body>
<h5 class="card-header" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 0px; font-weight: 500; line-height: 1.2; font-size: 1.25rem; padding: 1rem; color: rgb(52, 58, 64); border-bottom: none; border-radius: calc(0.375rem - 1px) calc(0.375rem - 1px) 0px 0px; font-family: system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, &quot;Helvetica Neue&quot;, Arial, &quot;Noto Sans&quot;, &quot;Liberation Sans&quot;, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Script Log</h5><div class="card-body" style="box-sizing: border-box; flex: 1 1 auto; padding: 0px 1rem 1rem; color: rgb(52, 58, 64); font-family: system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, &quot;Helvetica Neue&quot;, Arial, &quot;Noto Sans&quot;, &quot;Liberation Sans&quot;, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Line | Level | Message
-- | -- | --
1 | Info | updating testvm
2 | Info | field : None

</div>
</body>
</html>

```
            vms = VirtualMachine.objects.filter(name=vmName)
            if vms.count() == 1:
                if (not vms[0].custom_field_data["updated"]):
                    self.log_info("updating "+ vmName)
                    vms[0].custom_field_data["updated"] = True
                    vms[0].full_clean()
                    vms[0].save()
                    self.log_info("field :" + str(vms[0].custom_field_data["updated"]))

                else:
                    self.log_info("update is already True")

            elif vms.count() == 0:
                self.log_info("0 vm's found ignoring " + vmName)
            else:
                self.log_info("More than 1 vm found ignoring " +vmName)
```
I do get a change log update but nothing is changed. the custom field updated is still None. 
So I detect it needs to update the field but vms[0].save() does nothing?

**Comment by candlerb** [comment, marked as answer]
Just to check the obvious: you have created a Custom Field definition called "updated" in the Netbox UI, and set its data type to "boolean"? And you've associated this Custom Field definition with content type "Virtualization > virtual machine" ?

(Although if you hadn't done all that, the "clean" operation should have raised an exception)

And what Netbox version are you running?

**Reply by candlerb** [reply]
First I had to fill in the missing bits to make a runnable script:

```
from virtualization.models import VirtualMachine
from extras.scripts import Script, StringVar

class TestVM(Script):
    class Meta:
        name = "Test VM"
        description = "Test VM custom field"

    vm_name = StringVar(label="VM name")

    def run(self, data, commit):
            vmName = data["vm_name"]
            vms = VirtualMachine.objects.filter(name=vmName)
            if vms.count() == 1:
                if (not vms[0].custom_field_data.get("updated")):
                    self.log_info(f'updating {vmName}')
                    self.log_info(f'field : {vms[0].custom_field_data.get("updated")}')
                    vms[0].custom_field_data["updated"] = True
                    self.log_info(f'field : {vms[0].custom_field_data.get("updated")}')    # NOTE THIS!
                    vms[0].full_clean()
                    vms[0].save()
                    self.log_info(f'field : {vms[0].custom_field_data.get("updated")}')

                else:
                    self.log_info("updated is already True")

            elif vms.count() == 0:
                self.log_info("0 vm's found ignoring " + vmName)
            else:
                self.log_info("More than 1 vm found ignoring " +vmName)
```

What's clear from this is that Django fetches a fresh dict every time.  See the output of "NOTE THIS!"

The problem appears to be: `vms` is a proxy object (a query set), and every time you write `vms[0]` you are asking it to fetch a fresh copy of the object from the database.

To get around this, you can crystallize the query into real local objects, which you can do with `list(...)`

```
    def run(self, data, commit):
            vmName = data["vm_name"]
            vms = list(VirtualMachine.objects.filter(name=vmName))
            if len(vms) == 1:
                if (not vms[0].custom_field_data.get("updated")):
                    self.log_info(f'updating {vmName}')
                    self.log_info(f'field : {vms[0].custom_field_data.get("updated")}')
                    vms[0].custom_field_data["updated"] = True
                    self.log_info(f'field : {vms[0].custom_field_data.get("updated")}')
                    vms[0].full_clean()
                    vms[0].save()
                    self.log_info(f'field : {vms[0].custom_field_data.get("updated")}')

                else:
                    self.log_info("updated is already True")

            elif len(vms) == 0:
                self.log_info("0 vm's found ignoring " + vmName)
            else:
                self.log_info("More than 1 vm found ignoring " +vmName)
```

Or maybe just `vm = vms[0]` would be sufficient.

However, I think it would be better to simplify this script using [get()](https://docs.djangoproject.com/en/4.0/topics/db/queries/#retrieving-a-single-object-with-get):

```
vm = VirtualMachine.objects.get(name=vmName)
```

(which will give an exception if zero or multiple objects match). Even better, use Netbox's ObjectVar, which will allow the user to select a VM interactively from a drop-down list.

**Reply by MalfuncEddie** [reply]
Thank you so much!

**Reply by MalfuncEddie** [reply]
> Even better, use Netbox's ObjectVar, which will allow the user to select a VM interactively from a drop-down list.

Well actually it is a textfield where there is a multiline list of vm's that need to be updated. But since the names from 3rd party sources not necessarily match I wrote a method that first tries to find it by uuid, then by name or tenant ... (yes a lot of regex/filters
... why do you ask ;))

**Comment by MalfuncEddie** [comment]
Yes I added it, I also did a test with a vm that has (via GUI) the value True . Then I go to the self.log_info("update is already True") part
So it is a "saving" issue. Am I correct by your remark that this should work?

version v3.2.2

**Reply by MalfuncEddie** [reply]
I do get a 

Warning: Comparing non-atomic change to previous change record

In the changelog ( Difference: No Changes)
